cpp/meta_module/algorithm/meta.cpp

```diff
@@ -0,0 +1,330 @@
+          Meta::metadata.UpdateRelationshipTypes(edge, add, true);
+          Meta::metadata.UpdateRelationshipTypesCnt(edge, add, true);
+          Meta::metadata.UpdatePropertyKeyCnt(edge, add, true);
+        } else {
+          throw mgp::ValueException("Unexpected event type");
+        }
+      }
+    };
+
+    UpdateObject(created_objects, "created_vertex", "created_edge", 1);
+    UpdateObject(deleted_objects, "deleted_vertex", "deleted_edge", -1);
+
+    for (const auto &object : removed_vertex_properties) {
+      const auto event{object.ValueMap()};
+      LockAndExecute(metadata_mutex.property_key_cnt_mutex, true, [&]() {
+        auto key = std::string(event["key"].ValueString());
+        Insert(Meta::metadata.property_key_cnt, key, -1);
+      });
+    }
+
+    for (const auto &object : removed_edge_properties) {
+      const auto event{object.ValueMap()};
+      LockAndExecute(metadata_mutex.property_key_cnt_mutex, true, [&]() {
+        auto key = std::string(event["key"].ValueString());
+        Insert(Meta::metadata.property_key_cnt, key, -1);
+      });
+    }
+
+    for (const auto &object : set_vertex_labels) {
+      const auto event{object.ValueMap()};
+      LockAndExecute(metadata_mutex.property_key_cnt_mutex, true, [&]() {
+        auto key = std::string(event["label"].ValueString());
+        Insert(Meta::metadata.labels, key, static_cast<int64_t>(event["vertices"].ValueList().Size()));
+      });
+    }
+
+    for (const auto &object : removed_vertex_labels) {
+      const auto event{object.ValueMap()};
+      LockAndExecute(metadata_mutex.property_key_cnt_mutex, true, [&]() {
+        auto key = std::string(event["label"].ValueString());
+        Insert(Meta::metadata.labels, key, static_cast<int64_t>(-event["vertices"].ValueList().Size()));
+      });
+    }
+
+  } catch (const std::exception &e) {
+    record_factory.SetErrorMessage(e.what());
+    return;
+  }
+}
+
+void StatsOnline(mgp_list *args, mgp_graph *memgraph_graph, mgp_result *result, mgp_memory *memory) {
+  mgp::memory = memory;
+  const auto arguments = mgp::List(args);
+  const auto record_factory = mgp::RecordFactory(result);
+
+  try {
+    bool update_stats = arguments[0].ValueBool();
+
+    if (update_stats) {
+      Meta::metadata.Reset();
+      UpdateAllMetadata(memgraph_graph, metadata, true);
+      OutputFromMetadata(metadata, record_factory, true);
+    } else {
+      OutputFromMetadata(metadata, record_factory, true);
+    }
+
+  } catch (const std::exception &e) {
+    record_factory.SetErrorMessage(e.what());
+    return;
+  }
+}
+
+void StatsOffline(mgp_list *args, mgp_graph *memgraph_graph, mgp_result *result, mgp_memory *memory) {
+  mgp::memory = memory;
+  const auto arguments = mgp::List(args);
+  const auto record_factory = mgp::RecordFactory(result);
+
+  try {
+    Metadata data;
+    UpdateAllMetadata(memgraph_graph, data, false);
+    OutputFromMetadata(data, record_factory, false);
```

Comment:
I think `StatsOffline` should also reset metadata because of the case that users have already run `stats_online` before and saved metadata thus exists.

Reply:
If you have already ran the online version there is no reason to run the offline reason because the online version works in O(1). Also offline version doesn't modify the global metadata object.